File: Sigmath/Sigmath/Sigmath.ConvertedToVBNet/statistics/algorithms.vb
Imports System.Collections.Generic
Imports System.Linq

Public Class Statgorithm
	Public Shared Function Skewness(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate skewness of empty sequence")
		End If

		Dim count = 0
		Dim sum = 0.0
		Dim sumOfSquaredDifferences = 0.0
		Dim sumOfCubedDifferences = 0.0

		For Each value As var In values
			count += 1
			sum += value
			Dim difference = value - Statistics.Mean(values)
			sumOfSquaredDifferences += difference * difference
			sumOfCubedDifferences += difference * difference * difference
		Next

		Dim variance = Statistics.Variance(values)
		Dim skewness__1 = sumOfCubedDifferences / (count * Basic.CubeRoot(variance))

		Return skewness__1
	End Function

	Public Shared Function Kurtosis(values As IEnumerable(Of Double)) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate kurtosis of empty sequence")
		End If

		Dim mean = Statistics.Mean(values)
		Dim count = 0
		Dim sumOfSquaredDifferences = 0.0
		Dim sumOfFourthPowersOfDifferences = 0.0

		For Each value As var In values
			count += 1
			Dim difference = value - mean
			sumOfSquaredDifferences += difference * difference
			sumOfFourthPowersOfDifferences += difference * difference * difference * difference
		Next

		Dim variance = sumOfSquaredDifferences / count
		Dim fourthMoment = sumOfFourthPowersOfDifferences / count
		Dim kurtosis__1 = fourthMoment / (variance * variance) - 3

		Return kurtosis__1
	End Function

	Public Shared Function Covariance(xValues As IEnumerable(Of Double), yValues As IEnumerable(Of Double)) As Double
		If xValues Is Nothing OrElse Not xValues.Any() OrElse yValues Is Nothing OrElse Not yValues.Any() OrElse xValues.Count() <> yValues.Count() Then
			Throw New ArgumentException("Invalid input")
		End If

		Dim xMean = Statistics.Mean(xValues)
		Dim yMean = Statistics.Mean(yValues)
		Dim count = 0
		Dim sumOfProductsOfDifferences = 0.0

		Using xEnumerator = xValues.GetEnumerator()
			Using yEnumerator = yValues.GetEnumerator()
				While xEnumerator.MoveNext() AndAlso yEnumerator.MoveNext()
					count += 1
					Dim xDifference = xEnumerator.Current - xMean
					Dim yDifference = yEnumerator.Current - yMean
					sumOfProductsOfDifferences += xDifference * yDifference
				End While
			End Using
		End Using

		Dim covariance__1 = sumOfProductsOfDifferences / count

		Return covariance__1
	End Function



	Public Shared Function Percentile(values As IEnumerable(Of Double), percentile__1 As Integer) As Double
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot calculate percentile of empty sequence")
		End If

		If percentile__1 < 0 OrElse percentile__1 > 100 Then
			Throw New ArgumentOutOfRangeException("percentile", "Percentile must be between 0 and 100")
		End If

		Dim sortedValues = values.OrderBy(Function(v) v).ToArray()
		Dim index = CInt(Math.Truncate(Math.Ceiling((percentile__1 / 100.0) * sortedValues.Length))) - 1

		Return sortedValues(index)
	End Function

	Public Shared Function RemoveOutliers(values As IEnumerable(Of Double)) As IEnumerable(Of Double)
		If values Is Nothing OrElse Not values.Any() Then
			Throw New ArgumentException("Cannot remove outliers from empty sequence")
		End If

		Dim sortedValues = values.OrderBy(Function(v) v).ToArray()
		Dim q1 = Percentile(sortedValues, 25)
		Dim q3 = Percentile(sortedValues, 75)
		Dim iqr = q3 - q1
		Dim lowerBound = q1 - (1.5 * iqr)
		Dim upperBound = q3 + (1.5 * iqr)

		Return sortedValues.Where(Function(v) v >= lowerBound AndAlso v <= upperBound)
	End Function

End Class

File: statistics/algorithms.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Mathian
{
    public class Statgorithm
    {
        public static double Skewness(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate skewness of empty sequence");
            }

            var count = 0;
            var sum = 0.0;
            var sumOfSquaredDifferences = 0.0;
            var sumOfCubedDifferences = 0.0;

            foreach (var value in values)
            {
                count++;
                sum += value;
                var difference = value - Statistics.Mean(values);
                sumOfSquaredDifferences += difference * difference;
                sumOfCubedDifferences += difference * difference * difference;
            }

            var variance = Statistics.Variance(values);
            var skewness = sumOfCubedDifferences / (count * Basic.CubeRoot(variance));

            return skewness;
        }

        public static double Kurtosis(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate kurtosis of empty sequence");
            }

            var mean = Statistics.Mean(values);
            var count = 0;
            var sumOfSquaredDifferences = 0.0;
            var sumOfFourthPowersOfDifferences = 0.0;

            foreach (var value in values)
            {
                count++;
                var difference = value - mean;
                sumOfSquaredDifferences += difference * difference;
                sumOfFourthPowersOfDifferences += difference * difference * difference * difference;
            }

            var variance = sumOfSquaredDifferences / count;
            var fourthMoment = sumOfFourthPowersOfDifferences / count;
            var kurtosis = fourthMoment / (variance * variance) - 3;

            return kurtosis;
        }

        public static double Covariance(IEnumerable<double> xValues, IEnumerable<double> yValues)
        {
            if (xValues == null || !xValues.Any() || yValues == null || !yValues.Any() || xValues.Count() != yValues.Count())
            {
                throw new ArgumentException("Invalid input");
            }

            var xMean = Statistics.Mean(xValues);
            var yMean = Statistics.Mean(yValues);
            var count = 0;
            var sumOfProductsOfDifferences = 0.0;

            using (var xEnumerator = xValues.GetEnumerator())
            using (var yEnumerator = yValues.GetEnumerator())
            {
                while (xEnumerator.MoveNext() && yEnumerator.MoveNext())
                {
                    count++;
                    var xDifference = xEnumerator.Current - xMean;
                    var yDifference = yEnumerator.Current - yMean;
                    sumOfProductsOfDifferences += xDifference * yDifference;
                }
            }

            var covariance = sumOfProductsOfDifferences / count;

            return covariance;
        }

    

        public static double Percentile(IEnumerable<double> values, int percentile)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot calculate percentile of empty sequence");
            }

            if (percentile < 0 || percentile > 100)
            {
                throw new ArgumentOutOfRangeException("percentile", "Percentile must be between 0 and 100");
            }

            var sortedValues = values.OrderBy(v => v).ToArray();
            var index = (int)Math.Ceiling((percentile / 100.0) * sortedValues.Length) - 1;

            return sortedValues[index];
        }

        public static IEnumerable<double> RemoveOutliers(IEnumerable<double> values)
        {
            if (values == null || !values.Any())
            {
                throw new ArgumentException("Cannot remove outliers from empty sequence");
            }

            var sortedValues = values.OrderBy(v => v).ToArray();
            var q1 = Percentile(sortedValues, 25);
            var q3 = Percentile(sortedValues, 75);
            var iqr = q3 - q1;
            var lowerBound = q1 - (1.5 * iqr);
            var upperBound = q3 + (1.5 * iqr);

            return sortedValues.Where(v => v >= lowerBound && v <= upperBound);
        }

    }

}